pysynchronizer/icbinn.py

```diff
@@ -0,0 +1,241 @@
+ICBINN_LTYPE_RDLCK = 0
+ICBINN_LTYPE_WRLCK = 1
+ICBINN_LTYPE_UNLCK = 2
+ICBINN_RANDOM = 0
+ICBINN_URANDOM = 1
+
+DISK_DIR = 'disks'
+ENCRYPT_SNAPSHOTS = 'encrypt_snapshots'
+STATUS_OKAY = 0
+STATUS_INTERNAL_EXCEPTION = 1
+STATUS_FAILED = 2
+DOWNLOAD_BLOCK_SIZE = 512 * 1024
+ENCRYPTION_KEY_BYTES = 64
+PROGRESS_INTERVAL = 1
+DISK_TYPE_ISO = 'iso'
+DISK_TYPE_VHD = 'vhd'
+DISK_TYPES = [DISK_TYPE_ISO, DISK_TYPE_VHD]
+SYNC_ROLE_PLATFORM = 'platform'
+SYNC_ROLE_REALM = 'realm'
+SYNC_ROLES = [SYNC_ROLE_PLATFORM, SYNC_ROLE_REALM]
+RPC_PREFIX = 'rpc:'
+
+class Icbinn(Singleton):
+    def __init__(self):
+        uuid_node = Path('/sys/hypervisor/uuid')
+        if uuid_node.exists():
+            output = "/vm/" + uuid_node.read_text()
+        else:
+            output = Xenstore.read('vm').rstrip()
+
+        syncvm_object_path = output.replace('-', '_')
+        paths = OXTDBusApi.open_vm(syncvm_object_path).get_property('icbinn-path').split(',')
+
+        if len(paths) != 2:
+            raise InsufficientIcbinnPaths(2, paths)
+
+        self.storage = IcbinnRemote(paths[0], server_port=ICBINN_SERVER_PORT + 0)
+        self.config = IcbinnRemote(paths[1], server_port=ICBINN_SERVER_PORT + 1)
+
+class IcbinnRemote(object):
+    def __init__(self, mount_point, 
+                 server_domain_id=ICBINN_SERVER_DOMAIN_ID,
+                 server_port=ICBINN_SERVER_PORT):
+        self.mount_point = mount_point
+
+        try:
+            self.icbinn = icbinn_clnt_create_argo(server_domain_id,
+                                                 server_port)
+        except Exception as exc:
+            raise IcbinnConnectError("failed to connect to icbinn server at "
+                                     "(%d, %d): icbinn_clnt_create_argo "
+                                     "failed: %r" % (server_domain_id,
+                                                     server_port, exc))
+        if self.icbinn is None:
+            raise IcbinnConnectError("failed to connect to icbinn server at "
+                                     "(%d, %d): icbinn_clnt_create_argo "
+                                     "failed" % (server_domain_id,
+                                                 server_port))
+
+    def exists(self, path):
+        try:
+            res = self.stat(path)
+        except IcbinnError:
+            return False
+        return res[1] in [ICBINN_FILE, ICBINN_DIRECTORY]
+
+    def listdir(self, path):
+        files = []
+        while True:
+            entry = icbinn_readent(self.icbinn, str(path), len(files))
+            if entry is None:
+                break
+            files.append(entry[0])
+        return files
+
+    def mkdir(self, path):
+        if icbinn_mkdir(self.icbinn, str(path)) < 0:
+            raise PlatformError("error creating icbinn directory '%s': "
+                                "icbinn_mkdir failed" % path)
+
+    def makedirs(self, path, timeout=10):
+        # avoid leading and trailing slash confusion
```

Comment:
Maybe expand this comment with more detail.  You are avoiding leading slashes?

Reply:
Original code, so did not know the reasoning. I can see about trying to figure it out and document why.